finally updating the rate limits docs to be more helpful (#206)
Finch-API/finch-docs: docs/Development-Guides/Rate-Limits.md

# Rate Limits

Finch will return a rate limit error with the HTTP status code 429 when the request rate limit for an application, access token, or an individual IP address has been exceeded.

Finch's rate limits work on a per-product basis for both applications and access tokens. Rate limits are summed on a rolling 60-second basis for each unique `product`. This is commonly referred to as a Sliding or Rolling Window rate limit.

> You can think of a `product` rate limit like a "bucket". Therefore, when a request is made to a `product` (which correspond directly to an API endpoint), a single gallon of water is added to that endpoint’s bucket, thus starting that bucket's 60-second Time-To-Live (TTL) timer.

Reviewer: ```suggestion
You can think of a `product` rate limit like a "bucket". Therefore, when a request is made to a `product` (which corresponds directly to an API endpoint), a single gallon of water is added to that endpoint’s bucket, thus starting that bucket's 60-second time-to-live (TTL) timer.
```

Reviewer: nit: tokens or balls in a bucket might be more clear

> After the product's rate limit is reset after 60 seconds, the first request to that `product` starts the 60-second Time-To-Live (TTL) timer again.

Reviewer: ```suggestion
After the product's rate limit is reset after 60 seconds, the first request to that `product` starts the 60-second TTL again.
```
nit: after you define the acronym once you can just use it thereafter

***

## Access Token Rate Limits

> For access token rate limits, each `product` manages its own "bucket" with its own 60-second Time-To-Live (TTL) timer. The first request by an access token to a new product endpoint starts the timer for only that product's bucket, and rate limits reset only for that product endpoint after the 60-second period.

Reviewer: ```suggestion
For access token rate limits, each product has its own "bucket" with its own 60-second TTL timer. The first request made by an access token to a new product endpoint starts the timer for that specific product's bucket. Rate limits reset only for that product endpoint after the 60-second period.
```

`employment` | 4
`payment` | 2
`pay-statement` | 2

## Application Rate Limits

> Similarly, multiple access tokens can be created from a single Finch application as more employers are connected. A Finch application has its own rate limits separate from the access token rate limits. You can think of an application like a development environment or application stage. Finch will assign three types of applications: `sandbox`, `development`, and `production`. Each will have their own `client_id` which represents the application.

Reviewer 2: I'd omit the part about the kinds/types of applications, seems irrelevant to the rate limiting. I'd hope a developer knows their Client ID. I think that's a fairly common pattern.

Author: Unfortunately, some developers need this explained, so I'll keep it unless anyone else opposes it.

Reviewer 2: Maybe it should go in its own document/page explaining the Client ID differences?

For application-level rate limits, each `product` manages another “larger” bucket simultaneously counting all requests across all access tokens created by that application. Each bucket is still scoped to a product endpoint like with access tokens; the only difference is that the application-level bucket is larger.

Product | Max requests initiated per minute
-------|-------------
`company` | 20
`directory` | 20
`individual` | 20
`employment` | 20
`payment` | 12
`pay-statement` | 12

## IP Address Rate Limits

Finch also enforces individual IP Address rate limits. An IP address is a unique address that identifies a device on the internet or a local network. If the number of `max requests` are sent from the same IP Address within the `duration` time set, a `penalty` is enforced. No more requests are allowed once the penalty is enforced. Once the penalty duration is complete, requests will be accepted again.

Type | Max requests | Duration | Penalty
-------|-------------|-------|-------------
`API` | 1000 | 5 minutes | 60 minutes

## Other types of rate limits

### Finch Connect Rate limits

> Finch enforces rate limits on Employers going thru [Finch Connect](/docs/Integrating-with-Finch/Integrate-Finch-Connect/Overview.md). Each "step" in Finch Connect (the provider selection, logging in, multi-factor authentication prompts, etc) makes a single request to the Finch `/auth/authorize` endpoint. It is highly unlikely for an Employer to encounter these rate limit under normal use. However, an example of this rate limit being encountered is if an employer forgot their password and tried 30 different passwords in the time period of 1 minute. In this case, they would receive a rate limit penalty of 5 minutes.

Reviewer 2: Our Finch Connect React app (in Netlify) is the only client of `/auth/authorize` -- an external developer would never call this directly. It would make sense to omit this detail here (just adds confusion).

Author: Removed

Type | Max requests | Duration | Penalty
-------|-------------|-------|-------------
`Auth` | 30 | 1 minute | 5 minutes

### Upstream Provider Rate Limits

> Because Finch integrates directly with 180+ providers, we take extreme care to avoid provider-specific rate limits. However, it inevitably happens. In this case, we will return a specific error code [upstream_rate_limit_exceeded](/docs/Development-Guides/Errors/Error-Types.md#error-types-1) explaining that Finch encountered a provider-specific rate limit. In these cases, wait a few moments before retrying. It is difficult to recommend a wait duration since every provider is different.

Reviewer 2: No action item here -- but wanted to note this is an added benefit of getting entirely off of supporting "live" requests. These 429s are encountered by our job which has a greater "patience" to be able to slow up and retry. Even if our job can't break free of the rate-limiting, it would just end up as a failed job instead of being an error handled by the developer's code.

Author: Removed

## Scenarios

### Scenario 1: Hitting access token level rate limits

In this scenario, let's focus on a single access token (Token A) and assume you are making API requests to any of the the `company`, `directory`, `individual`, `employment`, `payment`, and `pay-statement` endpoints. When a request  is sent to an endpoint, a single gallon of water is added to the endpoint’s bucket each time. The first requests starts that bucket’s 60-second Time-To-Live (TTL) timer.

> 1. Token A makes `4` requests to the `directory` endpoint within a minute pouring 4 gallons of water into the “directory” bucket.

Reviewer 2: Instead of "`4`" I'd write out 4 and reserve the code fence for actual code (same for the other numbers below.

> 1. Token A makes `4` requests to the `directory` endpoint within a minute pouring 4 gallons of water into the “directory” bucket.

Reviewer 2: Instead of the shorthand "directory" I'd give the long hand GET /employer/directory (or whatever the path is)... also you could link to those docs ;)

Author: Linked to our API reference + updated links to /employer/directory

        `pay-statement` | 0/12
1. If Token F makes `1` request to the `company` endpoint and `1` request to the `directory` endpoint. The `company` request will fail and return a 429 rate limit error even though Token F has not maxed out its access-token specific `company` bucket. That is because the application-level bucket is now full. No requests from any token will succeed calling the `company` endpoint until the `company` bucket’s 60-second Time-To-Live (TTL) timer resets. However, Token F’s request to the `directory` endpoint will succeed because its access-token specific `directory` bucket is not full AND the application-level `directory` bucket is not full either.
    - Access Token F rate limits
        Bucket | Capacity
        -------|-------------
        `company` | 1/4
        `directory` | 1/4
        `individual` | 0/4
        `employment` | 0/4
        `payment` | 0/2
        `pay-statement` | 0/2
    - Application-level rate limits
        Bucket | Capacity
        -------|-------------
        `company` | 20/20 (FULL)
        `directory` | 16/20
        `individual` | 0/20
        `employment` | 0/20
        `payment` | 12/12 (FULL)

>         `pay-statement` | 0/12

Reviewer 2: This is very long-winded and more of a step-by-step example rather than a doc. Should this instead go on a new page?

Reviewer 2: Also @joshuakrohn an image may really really help us convey this information here.

Author: I agree that an image would be better, but working with what we have at the moment.

Author: @logangingerich where do you think this could go instead?

Reviewer 3: I had been thinking about it being a best practice doc. Alternatively, if we don't want to keep piling guides up, we could but it out as a blog post so it's at least easily shareable whenever folks need something more in-depth

Author: Refactored. I removed the sections and added a "Handling Rate Limit Errors" Best Practice guide.

`employment` | 4
`payment` | 2
`pay-statement` | 2

If an access token rate limit is encountered, it will contain the `finch_code`: [finch_token_rl](/docs/Development-Guides/Errors/Error-Types.md#error-types-1) in the response body.

```json
// HTTP 429 response body for access token rate limit exceeded
{
    “statusCode”: 429,
    “status”: 429,
    “code”: 429,
    “message”: “Too many requests for token”,
    “name”: “rate_limit_exceeded_error”,
    “finch_code”: “finch_token_rl”
}
```

## Application Rate Limits

> Similarly, multiple access tokens can be created from a single Finch application as more employers are connected. A Finch application has its own rate limits separate from the access token rate limits. You can think of an application like a development environment or application stage. Finch will assign three types of applications: `sandbox`, `development`, and `production`. Each will have their own `client_id` which represents the application.

Reviewer: ```suggestion
Similarly, multiple access tokens can be created from a single Finch application as more employers are connected. A Finch application has its own rate limits separate from the access token rate limits. (A Finch application corresponds to a unique `client_id`. You may have several `client_id`s if you use a development or sandbox application in addition to production).
```
I think we can find a middle ground and just be more concrete/succinct here.

{
    "statusCode": 429,
    "status": 429,
    "code”: 429,
    "message": "Too many requests for token",
    "name": "rate_limit_exceeded_error",
    "finch_code": "finch_application_rl"
}
```

## IP Address Rate Limits

Finch also enforces individual IP Address rate limits. An IP address is a unique address that identifies a device on the internet or a local network. If the number of `max requests` are sent from the same IP Address within the `duration` time set, a `penalty` is enforced. No more requests are allowed once the penalty is enforced. Once the penalty duration is complete, requests will be accepted again.

Type | Max requests | Duration | Penalty
-------|-------------|-------|-------------
`API` | 1000 | 5 minutes | 60 minutes

If an IP Address rate limit is encountered, it will contain the `finch_code`: [finch_api_ip_rl](/docs/Development-Guides/Errors/Error-Types.md#error-types-1) in the response body.

```json
// HTTP 429 response body for application rate limit exceeded
{
    "statusCode": 429,
    "status": 429,
    "code”: 429,
    "message": "Too many requests for token",
    "name": "rate_limit_exceeded_error",
    "finch_code": "finch_api_ip_rl"
}
```

***

> 

Reviewer: do we want to call out anywhere here that in addition to these rate limits, we can also return 429s if an upstream provider rate limits us?
